File: packages/dk-data-client/python/dk_data_client/modules/publications.py
"""Publications / literature read surface.

Reads from `mol_silver.publications` (unified), `mol_silver.pubmed_articles`,
and the upstream Europe PMC shim for fallthrough.
"""

from __future__ import annotations

from typing import Any

from dk_data_client.modules._base import ModuleBase


class PublicationsModule(ModuleBase):
    async def search(
        self, query: str, *, limit: int = 25, since: str | None = None
    ) -> list[dict[str, Any]]:
        params: dict[str, Any] = {
            "select": "publication_id,title,doi,pmid,published_at,journal",
            "title": f"ilike.*{query}*",
            "order": "published_at.desc",
            "limit": str(limit),
        }
        if since:
            params["published_at"] = f"gte.{since}"
        data = await self._call(
            method="publications.search",
            path="/publications",
            args={"query": query, "limit": limit, "since": since},
            params=params,
        )
        return list(data) if isinstance(data, list) else []

    async def get_by_molecule(self, molecule_id: str) -> list[dict[str, Any]]:
        params = {
            "molecule_id": f"eq.{molecule_id}",
            "order": "published_at.desc",
            "limit": "100",
        }
        data = await self._call(
            method="publications.getByMolecule",
            path="/publications",
            args={"molecule_id": molecule_id},
            params=params,
        )
        return list(data) if isinstance(data, list) else []

    async def get_pubmed(self, pmid: str) -> dict[str, Any]:
        params = {"pmid": f"eq.{pmid}", "limit": "1"}
        rows = await self._call(
            method="publications.getPubMed",
            path="/pubmed_articles",
            args={"pmid": pmid},
            params=params,
        )
        return dict(rows[0]) if isinstance(rows, list) and rows else {}

    async def get_openalex(self, work_id: str) -> dict[str, Any]:
        params = {"openalex_id": f"eq.{work_id}", "limit": "1"}
        rows = await self._call(
            method="publications.getOpenAlex",
            path="/publications",
            args={"openalex_id": work_id},
            params=params,
        )
        return dict(rows[0]) if isinstance(rows, list) and rows else {}

File: packages/dk-data-client/typescript/src/modules/publications.ts
/** Publications / literature read surface. */

import { ModuleBase } from "./_base.js";

export class PublicationsModule extends ModuleBase {
  async search(
    query: string,
    opts: { limit?: number; since?: string } = {},
  ): Promise<Array<Record<string, unknown>>> {
    const params: Record<string, string> = {
      select: "publication_id,title,doi,pmid,published_at,journal",
      title: `ilike.*${query}*`,
      order: "published_at.desc",
      limit: String(opts.limit ?? 25),
    };
    if (opts.since) params.published_at = `gte.${opts.since}`;
    const data = (await this.call({
      method: "publications.search",
      path: "/publications",
      args: { query, limit: opts.limit ?? 25, since: opts.since },
      params,
    })) as Array<Record<string, unknown>>;
    return Array.isArray(data) ? data : [];
  }

  async getByMolecule(moleculeId: string): Promise<Array<Record<string, unknown>>> {
    const data = (await this.call({
      method: "publications.getByMolecule",
      path: "/publications",
      args: { moleculeId },
      params: {
        molecule_id: `eq.${moleculeId}`,
        order: "published_at.desc",
        limit: "100",
      },
    })) as Array<Record<string, unknown>>;
    return Array.isArray(data) ? data : [];
  }

  async getPubMed(pmid: string): Promise<Record<string, unknown>> {
    const rows = (await this.call({
      method: "publications.getPubMed",
      path: "/pubmed_articles",
      args: { pmid },
      params: { pmid: `eq.${pmid}`, limit: "1" },
    })) as Array<Record<string, unknown>>;
    return Array.isArray(rows) && rows.length > 0 ? rows[0]! : {};
  }

  async getOpenAlex(workId: string): Promise<Record<string, unknown>> {
    const rows = (await this.call({
      method: "publications.getOpenAlex",
      path: "/publications",
      args: { workId },
      params: { openalex_id: `eq.${workId}`, limit: "1" },
    })) as Array<Record<string, unknown>>;
    return Array.isArray(rows) && rows.length > 0 ? rows[0]! : {};
  }
}
